service/storage_service.cc

```diff
@@ -5377,6 +5378,96 @@ future<> storage_service::update_fence_version(token_metadata::version_t new_ver
+    }
+    if (leaving.empty()) {
+        throw std::runtime_error(format("No leaving replicas"));
+    }
+    if (leaving.size() > 1) {
+        throw std::runtime_error(format("More than one leaving replica"));
+    }
+    return *leaving.begin();
+}
+
+inet_address storage_service::host2ip(locator::host_id host) {
+    auto ip = _group0->address_map().find(raft::server_id(host.uuid()));
+    if (!ip) {
+        throw std::runtime_error(::format("Cannot map host {} to ip", host));
+    }
+    return *ip;
+}
+
+// Streams data from a given tablet replica to the pending tablet replica on this node.
+future<> storage_service::stream_tablet(locator::global_tablet_id tablet) {
+    // The coordinator does not execute global token metadata barrier before jumping to "streaming" stage, so we need
+    // a barrier here to see the token metadata which is at least as recent as that of the sender.
+    auto& raft_server = _group0->group0_server();
+    co_await raft_server.read_barrier(&_abort_source);
+
+    auto tm = _shared_token_metadata.get();
+    auto& tmap = tm->tablets().get_tablet_map(tablet.table);
+    auto* trinfo = tmap.get_tablet_transition_info(tablet.tablet);
+
+    // Check if the request is still valid.
+    // If there is mismatch, it means this streaming was canceled and the coordinator moved on.
+    if (!trinfo) {
+        throw std::runtime_error(format("No transition info for tablet {}", tablet));
+    }
+    if (trinfo->stage != locator::tablet_transition_stage::streaming) {
+        throw std::runtime_error(format("Tablet {} stage is not at streaming", tablet));
+    }
+    if (trinfo->pending_replica.host != tm->get_my_id()) {
+        throw std::runtime_error(format("Tablet {} has pending replica different than this one", tablet));
+    }
+
+    auto& tinfo = tmap.get_tablet_info(tablet.tablet);
+    auto range = tmap.get_token_range(tablet.tablet);
+    locator::tablet_replica leaving_replica = get_leaving_replica(tinfo, *trinfo);
+    if (leaving_replica.host == tm->get_my_id()) {
+        // The algorithm doesn't work with tablet migration within the same node because
+        // it assumes there is only one tablet replica, picked by the sharder, on local node.
+        throw std::runtime_error(format("Cannot stream within the same node, tablet: {}, shard {} -> {}",
+                                        tablet, leaving_replica.shard, trinfo->pending_replica.shard));
+    }
+    auto leaving_replica_ip = host2ip(leaving_replica.host);
+
+    auto& table = _db.local().find_column_family(tablet.table);
+    std::vector<sstring> tables = {table.schema()->cf_name()};
+    auto streamer = make_lw_shared<dht::range_streamer>(_db, _stream_manager, tm, _abort_source,
+           get_broadcast_address(), _sys_ks.local().local_dc_rack(),
+           "Tablet migration", streaming::stream_reason::tablet, std::move(tables));
+    streamer->add_source_filter(std::make_unique<dht::range_streamer::failure_detector_source_filter>(
```

Comment:
Is it correct to stream from any replica of this tablet?

I think we should be following similar logic as for vnode replica sets, in order to preserve read-your-write consistency, i.e.:
- when we decommission a node, if that node was a replica for some tablet, we must stream from this replica and not any other
- but if some replica is dead and we remove it, we can stream from any replica -- or better, perform repair.

The requirement being that if we perform a quorum write followed by a quorum read, the read should observe the write. (In general if write_cl + read_cl > num_replicas)

Reply:
We don't stream from any replica, we stream from the leaving replica:
```
    ranges_per_endpoint[leaving_replica_ip].emplace_back(range);
```

Reply:
Handling of removenode is not implemented yet.